### PR TITLE
Refactor/shared module

### PR DESCRIPTION
**Pull request by alexmoroianu**

### CODE REVIEW

File: common/shared/pom.xml [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+  </parent>
+
+  <artifactId>shared</artifactId>
+
+  <dependencies>
+    <dependency>
+      <groupId>org.opencwa</groupId>
+      <artifactId>protocols</artifactId>
+    </dependency>
+    <dependency>
```
Comment (FelixRottler) [inline common/shared/pom.xml:19]: I think a lot of these dependencies are not needed inside this project. Maybe we can clean this pom.xml up a little bit

File: common/shared/src/main/java/app/coronawarn/server/common/shared/exception/UnableToLoadFileException.java [hunk 2]
```diff
@@ -0,0 +1,17 @@
+
+
+package app.coronawarn.server.common.shared.exception;
+
+/**
+ * The file could not be loaded/parsed correctly.
+ */
+public class UnableToLoadFileException extends Exception {
```
Comment (FelixRottler) [inline common/shared/src/main/java/app/coronawarn/server/common/shared/exception/UnableToLoadFileException.java:8]: In the distribution service this class exists too ( and is actually used)

File: common/shared/src/main/java/app/coronawarn/server/common/shared/util/HashUtils.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+   * that implements the chosen {@link MessageDigestAlgorithms}.
+   *
+   * @param data - ByteString to be diggested
+   * @return - digested ByteString
+   */
+  public static byte[] byteStringDigest(ByteString data, MessageDigestAlgorithms algorithm) {
+    try {
+      return MessageDigest.getInstance(algorithm.getName()).digest(data.toByteArray());
+    } catch (NoSuchAlgorithmException e) {
+      // DO NOTHING
```
Comment (FelixRottler) [inline common/shared/src/main/java/app/coronawarn/server/common/shared/util/HashUtils.java:52]: Empty catch

File: common/shared/src/test/java/app/coronawarn/server/common/shared/util/HashUtilsTest.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+import org.junit.jupiter.api.Test;
+import java.security.NoSuchAlgorithmException;
+
+class HashUtilsTest {
+
+  public static final String TEST_STRING = "string";
+
+  @Test
+  void testMd5DigestAsHex() {
+    String hash = md5DigestAsHex(TEST_STRING);
```
Comment (FelixRottler) [inline common/shared/src/test/java/app/coronawarn/server/common/shared/util/HashUtilsTest.java:17]: Maybe we can make this test more useful. We could take an existing hex md5 hash (and its origin) and see if this function gives the same result.

File: common/shared/src/test/java/app/coronawarn/server/common/shared/util/HashUtilsTest.java [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+    assertThat(hash).hasSize(32);
+  }
+
+  @Test
+  void testRandomByteArrayData() {
+    byte[] hash = generateSecureRandomByteArrayData(16);
+    assertThat(hash).hasSize(16);
```
Comment (FelixRottler) [inline common/shared/src/test/java/app/coronawarn/server/common/shared/util/HashUtilsTest.java:24]: Maybe we can also refactor this test. The advantage of the secure random is that even if we use the same seed for two different random variables the result will be different

File: services/distribution/src/main/java/app/coronawarn/server/services/distribution/assembly/structure/directory/IndexDirectory.java [hunk 6]
```diff
@@ -2,9 +2,9 @@
 
 package app.coronawarn.server.services.distribution.assembly.structure.directory;
 
+import app.coronawarn.server.common.shared.collection.ImmutableStack;
+import app.coronawarn.server.common.shared.functional.Formatter;
 import app.coronawarn.server.services.distribution.assembly.structure.Writable;
-import app.coronawarn.server.services.distribution.assembly.structure.util.ImmutableStack;
-import app.coronawarn.server.services.distribution.assembly.structure.util.functional.Formatter;
 import app.coronawarn.server.services.distribution.assembly.structure.util.functional.WritableFunction;
```
Comment (FelixRottler) [inline services/distribution/src/main/java/app/coronawarn/server/services/distribution/assembly/structure/directory/IndexDirectory.java:8]: Is there a reason that WritableFunction is not extracted into the shared module?

File: services/submission/src/test/java/app/coronawarn/server/services/submission/checkins/EventCheckinsFacadeIT.java [hunk 7]
```diff
@@ -52,17 +41,6 @@ public void setup() {
     currentTimestamp = CheckinsDateSpecification.HOUR_SINCE_EPOCH_DERIVATION.apply(Instant.now().getEpochSecond());
   }
 
-  @Test
```
Comment (FelixRottler) [inline services/submission/src/test/java/app/coronawarn/server/services/submission/checkins/EventCheckinsFacadeIT.java:55]: Why is that Test removed?

**Review by FelixRottler — CHANGES_REQUESTED** [review]
see comments :)

File: services/distribution/src/main/java/app/coronawarn/server/services/distribution/assembly/structure/directory/IndexDirectory.java (hunk 6 above)
Comment (alexmoroianu) [reply services/distribution/src/main/java/app/coronawarn/server/services/distribution/assembly/structure/directory/IndexDirectory.java:8]: Hi Felix. Actually, the entire "structure" package would make sense to be moved into an "io" package into shared module but I would do this in the next iteration as there are dozens of files and would really complicate this PR. What do you think ?

File: services/submission/src/test/java/app/coronawarn/server/services/submission/checkins/EventCheckinsFacadeIT.java (hunk 7 above)
Comment (alexmoroianu) [reply services/submission/src/test/java/app/coronawarn/server/services/submission/checkins/EventCheckinsFacadeIT.java]: I reverted it back.

File: common/shared/pom.xml (hunk 1 above)
Comment (alexmoroianu) [reply common/shared/pom.xml]: Removed them.

File: common/shared/src/main/java/app/coronawarn/server/common/shared/util/HashUtils.java (hunk 3 above)
Comment (alexmoroianu) [reply common/shared/src/main/java/app/coronawarn/server/common/shared/util/HashUtils.java]: Added a logger.

File: common/shared/src/main/java/app/coronawarn/server/common/shared/exception/UnableToLoadFileException.java (hunk 2 above)
Comment (alexmoroianu) [reply common/shared/src/main/java/app/coronawarn/server/common/shared/exception/UnableToLoadFileException.java]: Yes, indeed. I removed it from distribution and referenced it from shared.

File: common/shared/src/test/java/app/coronawarn/server/common/shared/util/HashUtilsTest.java (hunk 4 above)
Comment (alexmoroianu) [reply common/shared/src/test/java/app/coronawarn/server/common/shared/util/HashUtilsTest.java]: Done.

**Review by ioangut — APPROVED** [review]

**Review by FelixRottler — APPROVED** [review]

**Pull request merged**